sdk/python/feast/infra/offline_stores/contrib/mssql_offline_store/mssql.py

```diff
@@ -0,0 +1,647 @@
+            on_demand_feature_views=None,
+        )
+
+    @staticmethod
+    @log_exceptions_and_usage(offline_store="mssql")
+    def pull_all_from_table_or_query(
+        config: RepoConfig,
+        data_source: DataSource,
+        join_key_columns: List[str],
+        feature_name_columns: List[str],
+        timestamp_field: str,
+        start_date: datetime,
+        end_date: datetime,
+    ) -> RetrievalJob:
+        assert type(data_source).__name__ == "MsSqlServerSource"
+        warnings.warn(
+            "The Azure Synapse + Azure SQL offline store is an experimental feature in alpha development. "
+            "Some functionality may still be unstable so functionality can change in the future.",
+            RuntimeWarning,
+        )
+        from_expression = data_source.get_table_query_string().replace("`", "")
+        timestamps = [timestamp_field]
+        field_string = ", ".join(join_key_columns + feature_name_columns + timestamps)
+
+        query = f"""
+            SELECT {field_string}
+            FROM (
+                SELECT {field_string}
+                FROM {from_expression}
+                WHERE {timestamp_field} BETWEEN TIMESTAMP '{start_date}' AND TIMESTAMP '{end_date}'
+            )
+            """
+        engine = make_engine(config.offline_store)
+
+        return MsSqlServerRetrievalJob(
+            query=query,
+            engine=engine,
+            config=config.offline_store,
+            full_feature_names=False,
+            on_demand_feature_views=None,
+        )
+
+    @staticmethod
+    @log_exceptions_and_usage(offline_store="mssql")
+    def get_historical_features(
+        config: RepoConfig,
+        feature_views: List[FeatureView],
+        feature_refs: List[str],
+        entity_df: Union[pandas.DataFrame, str],
+        registry: BaseRegistry,
+        project: str,
+        full_feature_names: bool = False,
+    ) -> RetrievalJob:
+        warnings.warn(
+            "The Azure Synapse + Azure SQL offline store is an experimental feature in alpha development. "
+            "Some functionality may still be unstable so functionality can change in the future.",
+            RuntimeWarning,
+        )
+
+        expected_join_keys = _get_join_keys(project, feature_views, registry)
+        assert isinstance(config.offline_store, MsSqlServerOfflineStoreConfig)
+        engine = make_engine(config.offline_store)
+        if isinstance(entity_df, pandas.DataFrame):
+            entity_df_event_timestamp_col = (
+                offline_utils.infer_event_timestamp_from_entity_df(
+                    dict(zip(list(entity_df.columns), list(entity_df.dtypes)))
+                )
+            )
+            entity_df[entity_df_event_timestamp_col] = pandas.to_datetime(
+                entity_df[entity_df_event_timestamp_col], utc=True
+            ).fillna(pandas.Timestamp.now())
+
+        # TODO: figure out how to deal with entity dataframes that are strings
```

Comment:
can you throw an exception here? e.g. some not implemented error so users know they can't pass a SQL string for MsSQL to generate the entity dataframe?